packages/server/utils/AzureDevOpsServerManager.ts

```diff
@@ -0,0 +1,53 @@
+import fetch from 'node-fetch'
+import AzureDevOpsManager from 'parabol-client/utils/AzureDevOpsManager'
+import makeAppURL from 'parabol-client/utils/makeAppURL'
+import appOrigin from '../appOrigin'
+import {authorizeOAuth2} from '../integrations/helpers/authorizeOAuth2'
+import {
+  OAuth2PkceAuthorizationParams,
+  OAuth2PkceRefreshAuthorizationParams
+} from '../integrations/OAuth2Manager'
+
+class AzureDevOpsServerManager extends AzureDevOpsManager {
+  fetch = fetch as any
+
+  static async init(code: string, codeVerifier: string) {
+    return AzureDevOpsServerManager.fetchToken({
+      grant_type: 'authorization_code',
+      code: code,
+      code_verifier: codeVerifier,
+      redirect_uri: makeAppURL(appOrigin, 'auth/ado')
+    })
+  }
+
+  static async refresh(refreshToken: string) {
+    return AzureDevOpsServerManager.fetchToken({
+      grant_type: 'refresh_token',
+      refresh_token: refreshToken,
+      scope: '499b84ac-1321-427f-aa17-267ca6975798/.default',
+      redirect_uri: 'http://localhost:8081/'
+    })
+  }
+
+  private static async fetchToken(
+    params: OAuth2PkceAuthorizationParams | OAuth2PkceRefreshAuthorizationParams
+  ) {
+    const body = {
+      ...params,
+      client_id: process.env.AZUREDEVOPS_CLIENT_ID!
+    }
+    const additonalHeaders = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+      // eslint-disable-next-line prettier/prettier
+      'Origin': 'http://localhost:8081'
+    }
+    const authUrl = `https://login.microsoftonline.com/${process.env.AZUREDEVOPS_TENANT}/oauth2/v2.0/token`
```

Comment:
Correct.  I think I follow.  I'll add the tenants to the IntegrationProvider table

For context:
In the Azure DevOps world, each organization would be a tenant and each tenant has multiple projects.  A user can be a member of multiple organizations.

User----n*---->Tenant-------n*------>Project--------n*------->Work Item.

There are multiple types of work items (user story, bug, task, etc).

I'll store each tenant in the IntegrationProvider table

Reply:
@Dschoordsch  - For integration provider data in the IntegrationProvider table, how are you currently populating that table in production?  I am assuming that you are working with your customers' admins to provider the needed info (tenant, client_id, etc).   We didn't see any ui component or anything under /admin (besides /admin/graphql directly) for populating that table.  Should we assume that there is going to be an admin that is populating this table with the needed info, or is there a UI component that we are missing?

Reply:
I made that change.  I have so design options for the hierarchy.  In Azure DevOps we can get all work items for a user, but that would require multiple rest calls:

1. Find all ADO organizations I am a member of
2. Find all projects in each organization
3. Find all work items of a particular type in that project

We also have some nesting in other places.  We could hide the rest calls/constuction of the types interesting to the Parabol client in the RestManager, or we could expose the hierarchy in the GraphQL schema and handle it in resolvers.  My guess is that you will want to have a consistent graphql schema for your calling clients, but wanted to get your thoughts on it.

Reply:
@Dschoordsch - We made the broader changes that you were looking for.  There is still work to do on extending the schema.  Also, the token refresh pattern we implemented was based off the AtlassianLoaders.  In the newer pattern, it looks like the token refresh is being handled behind the JiraRestManager by an external library.  I can move the token refresh to the rest api layer if you would prefer that.

Reply:
> [...] the token refresh pattern we implemented was based off the AtlassianLoaders. In the newer pattern, it looks like the token refresh is being handled behind the JiraRestManager by an external library. I can move the token refresh to the rest api layer if you would prefer that.

Jira Server token effectively does not expire (valid for ~5 years), so there is no refresh logic at all. Atlassian integration has also some client logic to refetch the query once the token expired [useFreshToken](https://github.com/ParabolInc/parabol/blob/90b642f04f65df862ca941df896c7b09c08bf6db/packages/client/modules/teamDashboard/components/ProviderRow/AtlassianProviderRow.tsx#L48+L61). That's quite fragile though.
 
I think a far better solution would be to check the validity of the token right before usage and refresh it if necessary right before the request. This would require some locking as this could be run from multiple processes simultaneously. [`RedisLock`](https://github.com/ParabolInc/parabol/blob/90b642f04f65df862ca941df896c7b09c08bf6db/packages/server/utils/RedisLock.ts) would be the right thing to use for locking in this case.

This can be handled in subsequent PRs.